optaplanner-benchmarks/optaplanner-perf-framework/src/main/java/org/jboss/qa/brms/performance/profiler/MemoryConsumptionProfiler.java

```diff
@@ -0,0 +1,151 @@
+
+public class MemoryConsumptionProfiler implements InternalProfiler {
+
+    private static final Logger logger = LoggerFactory.getLogger(MemoryConsumptionProfiler.class);
+
+    private static final long MAX_WAIT_MSEC = 20 * 1000;
+
+    private final AtomicBoolean proceedWithMetrics = new AtomicBoolean(true);
+
+    private GcCounter gcCounter;
+
+    public MemoryConsumptionProfiler() {
+        try {
+            gcCounter = new GcCounter();
+        } catch (IllegalStateException ex) {
+            logger.error("{0} deactivated due to initialization error.", getClass().getSimpleName(), ex);
+            proceedWithMetrics.set(false);
+        }
+    }
+
+    @Override
+    public void beforeIteration(BenchmarkParams benchmarkParams, IterationParams iterationParams) {
+        // nothing to do
+    }
+
+    @Override
+    public Collection<? extends Result> afterIteration(BenchmarkParams benchmarkParams,
+                                                       IterationParams iterationParams,
+                                                       IterationResult result) {
+        if (!proceedWithMetrics.get()) {
+            return Collections.EMPTY_LIST;
+        }
+
+        final UsedMemory usedMemory;
+        try {
+            usedMemory = getUsedMemoryAfterSettling();
+        } catch (IllegalStateException ex) {
+            logger.error("{0} deactivated due to error during metrics recording.", getClass().getSimpleName(), ex);
+            return Collections.EMPTY_LIST;
+        }
+        List<Result> results = new ArrayList<>();
+        double usedHeap = bytesToKB(usedMemory.usedHeapMemory);
+        double usedNonHeap = bytesToKB(usedMemory.usedNonHeapMemory);
+        double total = bytesToKB(usedMemory.totalCommittedMemory);
+        results.add(new ScalarResult("mem.used.heap", usedHeap, "kB", AggregationPolicy.AVG));
+        results.add(new ScalarResult("mem.used.nonheap", usedNonHeap, "kB", AggregationPolicy.AVG));
+        results.add(new ScalarResult("mem.total", total, "kB", AggregationPolicy.AVG));
+
+        return results;
+    }
+
+    private double bytesToKB(long bytes) {
+        return bytes / 1024;
+    }
+
+    private UsedMemory getUsedMemoryAfterGc() {
+        long beforeGcCount = gcCounter.count();
+        long startMsec = System.currentTimeMillis();
+
+        System.gc();
+        while (System.currentTimeMillis() - startMsec < MAX_WAIT_MSEC) {
+            try {
+                Thread.sleep(234);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+
+            if ((gcCounter.count() > beforeGcCount)) {
+                return getUsedMemorySimply();
+            }
+        }
+        throw new IllegalStateException("GS hasn't been detected after calling System.gc()");
+    }
+
+    private UsedMemory getUsedMemorySimply() {
+        MemoryUsage heapUsage = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
```

Comment:
Nitpicking: heapUsage and nonHeapUsage can be also final as the other local variables. Please check also the other part of this PR, so it is used everywhere or nowhere, so the usage of `final` is consistent.